woocommerce/woocommerce-gutenberg-products-block
# Handle change to logged-in status in checkout API AJAX requests

Related #3363 #3423

The checkout API can sometimes create a new user account and log that user in – the `Create an account?` option.

If there are other errors that prevent checkout from succeeding (e.g. invalid coupon - see #3363), this can leave the checkout in a slightly broken state. The core of the issue is that the checkout block and state isn't updated to reflect the fact that the user is logged in. Also the API response includes a stale nonce, because the nonce is generated early in request handling, before account creation.

The goal of this PR is to fix these issues, and generally make the checkout UX robust to user login each time the checkout is submitted (i.e. after any AJAX request).

- Ensure the response includes a valid nonce.
- Trigger changes to state/props so the UX updates appropriately. For example:
  - Hide the `Create an account?` checkbox if user is now logged in.
  - Show a notice informing the user that their account was created, even though the order submission failed.

This PR is currently based off #3454. They are closely related; this PR depends on the updated nonce from #3454 when a new account is created. When #3454 is merged, this should be rebased to trunk (and retested).

### How to test the changes in this Pull Request:

1. Make a coupon with an email restriction e.g. `*@legit.com`.
1. In logged-out browser/user, add a product to cart, proceed to checkout.
1. Enter coupon code.
1. Enter an invalid email for coupon e.g. `try@sneaky.com`
1. Check `Create an account?` option and complete checkout fields.
1. Submit checkout.
1. Confirm that checkout shows appropriate feedback:
  - Invalid coupon has been removed.
  - New user account created - `Create an account` checkbox is gone + notice.
1. Submit checkout again - should succeed.

### "New account" notice

Note - I couldn't get a regular notice working, it seems to only display one at a time. Snackbar style works, and maybe is a good fit for this notice. @LevinMedia can you review this notice UX and wording and advise? I don't think the notice is essential (could be handled as an improvement later), so we could back out of this if needed.

Here are some ideas for wording for the notice:

  - `Congratulations! You have signed up for a user account.`
  - `Welcome! Your user account has been created.`
  - variations of the above with email / username

### Screenshot walkthrough

![Screen Capture on 2020-11-27 at 14-05-21](https://user-images.githubusercontent.com/4167300/100399962-b34a3500-30b9-11eb-9c9c-c383fa0c6e1a.gif)


### Changelog

> Fix – Checkout block: Correctly handle cases where the order fails with an error (e.g. invalid coupon) and a new user account is created.
